File: logguard-js/logger.js
import * as fs from 'fs';
import * as path from 'path';

class FileNotOpen extends Error {} // FileNotOpen Error
class PathNonExistant extends Error {} // PathNonExistant Error


class Logger {
    /**
     * @param {string} [output_dir='logs'] - The directory where log files will be stored.
     * @param {string} [log_file_type='log'] - The type of log file (e.g., 'log', 'txt').
     * @param {string} [settings_path='./log_settings.json'] - The path to the log settings file.
     * @param {string} [LogLevel='INFO'] - The log level (e.g., 'INFO', 'DEBUG') to be ignored.
     * @throws {Error} If the log file type isnt supported.
     * @description  The logger class
     */
    
    constructor(output_dir = 'logs', log_file_type = 'log', settings_path = './log_settings.json', LogLevel = 'INFO') {
        // Set log level to upper case
        this.loglevel = LogLevel.toUpperCase();
        this.open_loggers = {}
        // Initialise variables
        this.settings = null;
        this.configured_level_value = null;
        this.file_path = null;
        this.log_file = null;
        this.log_file_name = null;
        this.timestamp = null;

        // Assign values
        // Set timestamp to current date and time in the specified format
        this.timestamp = new Date().toISOString().replace(/:/g, '-').replace(/T/, '_').replace(/\..+/, '');
        // List of supported log file formats
        this.supported_formats = ['log', 'txt'];
        // If settings are available, set configured log level value

        // Check log file type
        if (this.supported_formats.includes(log_file_type)) {
            // Set log file type
            this.log_file_type = log_file_type;
        } else {
            // Raise an error if log file type is not supported
            throw new Error('Log file type isn\'t supported');
        }

        // Check for the output directory
        if (output_dir === '.') {
            // If output directory is current directory, set file path to current working directory
            this.file_path = path.resolve();
        } else {
            // Otherwise, set file path to specified directory
            this.file_path = path.resolve(output_dir);
        }

        // Check the settings file path
        this.settings_path = settings_path;
        this.load_json(this.settings_path);

        if (this.settings) {
            this.configured_level_value = this.settings['LogLevels'][this.loglevel];
        }

        // Create the log file
        this.create_log_file(); // Ensure log file is created during initialization
    }

    /**
     * @param {string} level - The log level (e.g., 'INFO', 'DEBUG').
     * @param {string} message - The message to log.
     * @param {*} [context=null] - Additional context for the log message.
     * @throws {FileNotOpen} If the settings or the Log file isnt open.
     * @description Logs the message to the log file
     */
    log(level, message, context = null) {
        const timestamp = new Date().toISOString().replace(/:/g, '-').replace(/T/, '_').replace(/\..+/, '');
        level = level.toUpperCase(); // make the level upper case if it isn't

        if (this.log_file) { // if the log file exists
            if (this.settings) { // if the self.settings variable is not null
                const log_level_value = this.settings['LogLevels'][level];

                if (log_level_value && this.configured_level_value) {
                    // check that the configured log level is lower than the given log level
                    if (log_level_value >= this.configured_level_value) {
                        if (timestamp) {
                            // write the formatted message to the log file
                            if (context) {
                                const formatted_message = this.Formatter(level, message, timestamp, true, context);
                                if (formatted_message !== null) {
                                    this.log_file.write(formatted_message);
                                    
                                }
                            } else {
                                const formatted_message = this.Formatter(level, message, timestamp);
                                if (formatted_message !== null) {
                                    this.log_file.write(formatted_message);
                                    
                                }
                            }
                        }
                    }
                } else {
                    // Log level is lower than configured level, do nothing
                }
            } else {
                throw new FileNotOpen('Settings file is not open');
            }
        } else {
            throw new FileNotOpen('Log file is not open');
        }
    }

    /**
    * @param {string} [level] -The severity level.
    * @param {string} [message] - The message to be logged.
    * @param {string} [timestamp] - The time at which the message was logged.
    * @param {any} [context] - The context to be logged.
    * @throws {FileNotOpen} If the settings file is not open.
    * @description Format the message
    */
    Formatter(level, message, timestamp, context_value = false, context = null) {
        if (this.settings) {
            if (context_value) {
                if (context) {
                    const format_template = this.settings['Formats']['Context'];
                    const formatted_message = format_template.replace('{level}', level)
                        .replace('{message}', message)
                        .replace('{timestamp}', timestamp)
                        .replace('{context}', context);
                    return formatted_message + '\n';
                }
            } else {
                const format_template = this.settings['Formats']['NonContext'];
                const formatted_message = format_template
                    .replace('{level}', level)
                    .replace('{message}', message)
                    .replace('{timestamp}', timestamp);
                return formatted_message + '\n';
            }
        } else {
            throw new FileNotOpen('Settings file is not open');
        }
    }

    /**
     * @throws {PathNonExistant} If the the path is null.
     * @description Create the log file in the specified directory.
     */
    create_log_file() {
        this.log_file_name = this.get_log_name(); // get the log file name
    
        if (this.file_path) {
            fs.mkdirSync(this.file_path, { recursive: true }); // make the log dir if it doesn't exist
        } else {
            throw new PathNonExistant('File path not found'); // raise an error if file path is  null
        }
    
        try {
            if (this.log_file_name in this.open_loggers) { // check if there is an already open logger file
                this.log_file = this.open_loggers[this.log_file_name]; // open that logger file
            } else {
                if (!fs.existsSync(this.log_file_name)) { // check if log file exists
                    fs.writeFileSync(this.log_file_name, ''); // create the log file if it doesn't exist
                }
                this.log_file = fs.createWriteStream(this.log_file_name, { flags: 'a' }); // open the log file
                this.open_loggers[this.log_file_name] = this.log_file; // store the opened logger file
                this.log('info', 'Starting'); // set a starting message
            }
        } catch (error) {
           
            throw new Error(`Unexpected Error ${error}`); 
        }
    }
     
    /**
     * @throws {PathNonExistant} If file path is null
     * @description Get the full path of the log file
     */
    get_log_name() {
        if (this.file_path) {
            return path.resolve(this.file_path, `${this.timestamp}.${this.log_file_type}`); // return the filepath and the name of the file
        } else {
            throw new PathNonExistant('File path not found'); // raise an error if file path is  null
        }
    }

    /**
    * @param {string} [path] - The path to be opened
    * @throws {PathNonExistant} If the file path does not exist
    * @description Opens json files
    */
    load_json(path) {
        if (fs.existsSync(path)) {
            const data = fs.readFileSync(path, 'utf8');
            this.settings = JSON.parse(data);
        } else {
            throw new PathNonExistant('File path does not exist');
        }
    }
    
    /**
     *@throws {FileNotOpen} If the log file is not open 
     *@description Closes the current log file
     */
    close() {
        if (this.log_file) {
            try {
                this.log('info', 'Closing file');
            } finally {
                this.log_file.close();
                delete Logger.open_loggers[this.log_file_name];
                this.log_file = null;
            }
        } else {
            throw new FileNotOpen('Log file is not open');
        }
    }
}

export { Logger };

File: logguard-py/logger.py
import datetime
import sys
from pathlib import Path
import threading
import json

__all__ = ["Logger"]

# Custom Errors
class Errors:
    """Contains Custom Errors"""

    class FileNotOpen(Exception):
        """FileNotOpen Error"""
        

    class PathNonExistent(Exception):
        """PathNonExistent Error"""
        

    class LevelNotSupported(Exception):
        """LevelNotSupported Error"""

    class UnableToLock(Exception):
        """UnableToLock Error"""

    class LockNonExistent(Exception):
        """LockNonExistent Error"""

class Logger:
    """
        Initialize Logger instance.

        Args:
            output_dir (str, optional): the output location. Defaults to 'logs'.
            log_file_type (str, optional): the file log type. Defaults to 'log'.
            settings_path (str, optional): the settings path. Defaults to './log_settings.json'.
            log_level (str, optional): the log level. Defaults to 'INFO'.

        Raises:
            ValueError: If the log file type isn't supported.
            PathNonExistent: If the log file path does not exist.
            FileNotOpen: If the log file is not open.
            PathNonExistent: If the log settings file path does not exist.
            UnableToLock: If the script is unable to get a thread lock.
        """
    open_loggers_lock = threading.Lock()  # Lock for synchronizing access to open_loggers
    open_loggers = {}  # Thread-safe dictionary to store open loggers

    def __init__(self,
                 output_dir: str = "logs",
                 log_file_type: str = "log",
                 settings_path: str = './log_settings.json',
                 log_level: str = 'INFO'):
        """
        Initialize Logger instance.

        Args:
            output_dir (str, optional): the output location. Defaults to 'logs'.
            log_file_type (str, optional): the file log type. Defaults to 'log'.
            settings_path (str, optional): the settings path. Defaults to './log_settings.json'.
            log_level (str, optional): the log level. Defaults to 'INFO'.

        Raises:
            ValueError: If the log file type isn't supported.
            PathNonExistent: If the log file path does not exist.
            FileNotOpen: If the log file is not open.
            PathNonExistent: If the log settings file path does not exist.
            UnableToLock: If the script is unable to get a thread lock.
        """
        self.lock = None

        try:
            # Attempt to create a thread lock
            self.lock = threading.Lock()
        except Exception as e:
            # Raise an error if unable to create a thread lock
            raise Errors.UnableToLock(f"Unable to get process lock with error {e}")

        # Set log level to upper case
        self.log_level = log_level.upper()

        # Initialize variables
        self.settings = None
        self.configured_level_value = None
        self.file_path = None
        self.log_file = None
        self.log_file_name = None
        self.timestamp = None

        # Assign values
        self.timestamp = datetime.datetime.now().strftime("%Y-%m-%d_%H-%M-%S")
        self.supported_formats = ["log", "txt"]
        self.log_file_type = log_file_type if log_file_type in self.supported_formats else 'log'
        self.file_path = Path.cwd().resolve() if output_dir == "." else Path(output_dir).resolve()
        self.settings_path = settings_path
        self.load_json(self.settings_path)
        if self.settings:
            self.configured_level_value = self.settings['LogLevels'][self.log_level]

        # Create the log file
        self.create_log_file()

    def log(self, level: str, message: str, context=None):
        """
        Make a new log to the log file.

        Args:
            level (str): The log level.
            message (str): The log message.
            context (dict, optional): Contextual information to be included in the log message.

        Raises:
            FileNotOpen: If log file is not open.
            LockNonExistent: If the script doesn't have a lock.
            FileNotOpen: If the log file is not open.
        """
        timestamp = datetime.datetime.now().strftime("%Y-%m-%d-%H-%M-%S")
        level = level.upper()

        if self.lock:
            with self.lock:
                if self.log_file:
                    if self.settings:
                        log_level_value = self.settings['LogLevels'][level]
                        if log_level_value and self.configured_level_value:
                            if log_level_value >= self.configured_level_value:
                                if timestamp:
                                    if context:
                                        formatted_message = self.formatter(
                                                                        level, message,
                                                                           timestamp,
                                                                           context=context)
                                    else:
                                        formatted_message = self.formatter(level, message,
                                                                           str(timestamp))

                                    if formatted_message is not None:
                                        self.log_file.write(formatted_message)
                                        self.log_file.flush()
                            else:
                                pass
                    else:
                        raise Errors.FileNotOpen('Settings file is not open')
                else:
                    raise Errors.FileNotOpen('Log file is not open')
        else:
            raise Errors.LockNonExistent("Self.lock is None or false for some reason contact dev ")

    def formatter(self, level: str, message: str, timestamp: str, context: object = None):
        """
        Format log message.

        Args:
            level (str): The log level.
            message (str): The log message.
            timestamp (str): The timestamp.
            context (dict, optional): The context to add.

        Returns:
            str: Formatted log message.

        Raises:
            FileNotOpen: If the settings file is not open.
        """
        if self.settings:
            if context:
                format_template = self.settings['Formats']['Context']
                formatted_message = format_template.format(level=level, message=message,
                                                           timestamp=timestamp, context=context)
            else:
                format_template = self.settings['Formats']['NonContext']
                formatted_message = format_template.format(level=level, message=message
                                                           , timestamp=timestamp)
                
            return formatted_message + "\n"
        raise Errors.FileNotOpen('Settings file is not open')

    def create_log_file(self):
        """
        Generate the log file.

        Raises:
            PathNonExistent: If the file path is not found.
        """
        self.log_file_name = self.get_log_name()

        if self.file_path:
            self.file_path.mkdir(parents=True, exist_ok=True)
        else:
            raise Errors.PathNonExistent("File path not found")

        try:
            if self.log_file_name in Logger.open_loggers:
                self.log_file = Logger.open_loggers[self.log_file_name]
            else:
                with open(str(self.log_file_name), "a",encoding="utf-8") as log_file: 
                    self.log_file = log_file
                Logger.open_loggers[self.log_file_name] = self.log_file
                self.log('info', "Starting")
        except IOError:
            sys.stderr.write(f"Error: Unable to open log file {self.log_file_name}\n")

    def get_log_name(self):
        """
        Get the log file name.

        Returns:
            str: The log file path and the name of it.

        Raises:
            PathNonExistent: If file path is not found.
        """
        if self.file_path:
            return self.file_path / f"{self.timestamp}.{self.log_file_type}"
        raise Errors.PathNonExistent("File path not found")

    def load_json(self, path: str):
        """
        Load JSON settings.

        Args:
            path (str): The path to the JSON settings file.

        Raises:
            PathNonExistent: If the file path does not exist.
            LockNonExistent: If the script doesn't have a lock.
        """
        if self.lock:
            with self.lock:
                if Path(path).resolve():
                    with open(path, 'r',encoding="utf-8") as f:
                        self.settings = json.load(f)
                else:
                    raise Errors.PathNonExistent('File path does not exist')
        else:
            raise Errors.LockNonExistent('Thread lock does not exist, please contact dev')

    def close(self):
        """
        Close the current log file.

        Raises:
            FileNotOpen: If file is not open.
            LockNonExistent: If the script doesn't have a lock.
        """
        if self.log_file:
            try:
                self.log('info', "Closing file")
            finally:
                if self.lock:
                    with self.lock:
                        self.log_file.close()
                        with self.open_loggers_lock:
                            if self.log_file_name in self.open_loggers:
                                del self.open_loggers[self.log_file_name]
                        self.log_file = None
                        self.lock.release()
                else:
                    raise Errors.LockNonExistent('Threading lock is None or False for some reason')
        raise Errors.FileNotOpen("Log file is not open")
        